compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/ResolveUtils.kt

```diff
@@ -247,9 +247,35 @@ private fun BodyResolveComponents.typeFromSymbol(symbol: AbstractFirBasedSymbol<
     }
 }
 
-fun BodyResolveComponents.transformQualifiedAccessUsingSmartcastInfo(qualifiedAccessExpression: FirQualifiedAccessExpression): FirQualifiedAccessExpression {
+fun BodyResolveComponents.transformQualifiedAccessUsingSmartcastInfo(
+    qualifiedAccessExpression: FirQualifiedAccessExpression
+): FirQualifiedAccessExpression {
     val typesFromSmartCast = dataFlowAnalyzer.getTypeUsingSmartcastInfo(qualifiedAccessExpression) ?: return qualifiedAccessExpression
     val originalType = qualifiedAccessExpression.resultType.coneType
+    // For example, if (x == null) { ... },
+    //   we don't want to smartcast to Nothing?, but we want to record the nullability to its own kind of node.
+    // TODO: should we differentiate x == null v.s. x is Nothing?
+    if (typesFromSmartCast.any { it.isNullableNothing }) {
+        val typesFromSmartcastWithoutNullableNothing =
+            typesFromSmartCast.filterTo(mutableListOf()) { !it.isNullableNothing }.also {
+                it += originalType
```

Comment:
An alternative I tried was:
```kt
it += originalType.withNullability(...NULLABLE)
```
because, again, it's cast to literally `null`.

But, encountered other kind of non-sense cases, like,
```kt
fun foo(x : NonNull) {
  if (x == null) { ... } // ?!
}
```
It's better to see `<!DEBUG_INFO_EXPRESSION_TYPE(NonNull & NonNull)!><!>` (where the 2nd one is the original type) instead of `... NonNull & NonNull?`. So, again, ended up with the current form.